pkg/bytecode/code.go

```diff
@@ -0,0 +1,96 @@
+package bytecode
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+)
+
+const (
+	// OpConstant defines a constant that will be referred to by index
+	// in the bytecode.
+	OpConstant Opcode = iota
+	// OpGetGlobal retrieves a symbol from the symbol table at the
+	// specified index.
+	OpGetGlobal
+	// OpSetGlobal adds a symbol to the specified index in the symbol
+	// table.
+	OpSetGlobal
+)
+
+// ErrUnknownOp is returned when an unknown opcode is encountered.
+var ErrUnknownOp = errors.New("unknown opcode")
+
+// definitions is a mapping of OpCode to OpDefinition.
+var definitions = map[Opcode]*OpDefinition{
+	// The definition for OpConstant says that its only operand is two
+	// bytes wide, which makes it a uint16. This means that an evy
+	// bytecode program can only have 65535 constants defined.
+	OpConstant:  {"OpConstant", []int{2}},
+	OpGetGlobal: {"OpGetGlobal", []int{2}},
+	OpSetGlobal: {"OpSetGlobal", []int{2}},
+}
+
+// OpDefinition defines a name and expected operand width for each OpCode.
+type OpDefinition struct {
+	Name          string
+	OperandWidths []int
+}
+
+// Opcode defines the type of operation to be performed when reading
+// the bytecode.
+type Opcode byte
+
+// Lookup returns an OpDefinition for the value of an Opcode, any
+// unknown value will result in an error.
+func Lookup(op byte) (*OpDefinition, error) {
```

Comment:
This name is a little generic package-wide - you also have a symbol table which is something you look up. Perhaps call it `LookupOpcode()`?

Also, why does it take a `byte` and not an `Opcode`? In the three uses of this function, two have an Opcode and cast it to a byte. The other arguably should cast `ins[i]` to `Opcode`. I guess there's always going to be a little tension between when  you use the raw bytes as it comes from the `Instructions` and when you use the higher level types for what those bytes represent. This may need some more thought to keep these layers clearly separated and to keep like code grouped together.

Reply:
> This name is a little generic package-wide

I think it's fine as-is, symbol table uses Define/Resolve which is distinct enough from lookup.

> Also, why does it take a byte and not an Opcode?

On reflection I think it makes sense to change it to accept an Opcode.

Reply:
`Lookup` is package scoped and exported so when I see `bytecode.Lookup()` in code, I do not immediately consider that it is looking up Opcodes in the definitions map, given that the `bytecode` package contains symbol tables, symbols, vms, compilers, instructions, etc. If this were in a package `opcode`, it would be fine IMHO.

My point about symbol table was that is a type in the package that very clearly screams "lookup" - being a table. Of all the types in the package, `SymbolTable` is the one I would most closely associate with the bare word `Lookup`, but that was more of an example of how the vagueness of the name `Lookup` could cause confusion, not something specifically about `SymbolTable`.

But now that the function takes an `Opcode` and not a `byte`, it is a bit clearer as callers to `Lookup` will either pass a var probably named `op` / `opcode` or if they have a `byte`, it would need to be `Opcode(byte)`.

So if you'd prefer to keep it `Lookup`, no worries.